Sec5.2_validation/check_consistency.sage

```diff
@@ -83,27 +96,108 @@ def test_dbdd_ppred(n, m, q, D_e, D_s, nb_hints, hint_weight):
+    check_all_equal(dbdd, dbdd_o)
     check_all_equal(dbdd, dbdd_p)
 
     if hw == 1:
         dbdd_d.integrate_q_vectors(q)
         check_all_equal(dbdd, dbdd_d)
 
     dbdd.estimate_attack()
+    dbdd_o.estimate_attack()
     dbdd_p.estimate_attack()
     if hw == 1:
         dbdd_d.estimate_attack()
-    dbdd.attack()
+    _, solution = dbdd.attack()
+    _, solution_o = dbdd_o.attack()
+    assert (solution is not None) and (solution_o is not None)
+    logging("Passed that batch of tests !", style="SUCCESS")
+
+
+def test_dbdd_optimized(n, m, q, D_e, D_s, nb_hints, hint_weight):
+    A, b, dbdd = initialize_from_LWE_instance(DBDD, n, q, m,
+                                              D_e, D_s,
+                                              verbosity=2)
+    dbdd_o = DBDD_optimized(
+        copy(dbdd.B), copy(dbdd.S), copy(dbdd.mu), dbdd.u,
+        D=copy(dbdd.D), verbosity=2,
+    )
+    d = n + m
+    dbdd.estimate_attack()
+    dbdd_o.estimate_attack()
+    check_all_equal(dbdd, dbdd_o)
+    print("Ok so far.")
+
+    from time import time as get_time
+    get_duration = lambda t: (get_time() - t)
+    basic_time = 0
+    optimized_time = 0
+
+    for h in range(nb_hints):
+        v = vec(d * [0])
+        for w in range(hint_weight):
+            p = randint(0, d - 1)
+            v[0, p] = randint(- 1, 1) if w < hint_weight - 1 else 1
+
+        leak = dbdd.leak(v)
+        if (h % 3) == 0:
+            integrate_hint_method = "integrate_perfect_hint"
+            args, kwargs = (v, leak), {"estimate": False}
+
+        elif (h % 3) == 1:
+            k = randint(2, 12)
+            integrate_hint_method = "integrate_modular_hint"
+            args, kwargs = (v, leak % k, k), {"smooth": True, "estimate": False}
+
+        else:
+            sigma = randint(1, 10) / 10
+            apost = bool(randint(0, 1))
+            integrate_hint_method = "integrate_approx_hint"
+            args, kwargs = (v, leak, sigma), {"aposteriori": apost, "estimate": False}
+
+        # Integrate the hint for the DBDD object
+        time_point = get_time()
+        getattr(dbdd, integrate_hint_method)(*args, **kwargs)
+        basic_time += get_duration(time_point)
+
+        # Integrate the hint for the DBDD_optimized object
+        time_point = get_time()
+        getattr(dbdd_o, integrate_hint_method)(*args, **kwargs)
+        optimized_time += get_duration(time_point)
+
+    # Integrate the q-vectors for the DBDD object
+    time_point = get_time()
+    dbdd.integrate_q_vectors(q)
+    dbdd.estimate_attack()
+    basic_time += get_duration(time_point)
+
+    # Integrate the q-vectors for the DBDD_optimized object
+    time_point = get_time()
+    dbdd_o.integrate_q_vectors(q)
+    dbdd_o.estimate_attack()
+    optimized_time += get_duration(time_point)
+
+    check_all_equal(dbdd, dbdd_o)
+
+    _, solution = dbdd.attack()
+    _, solution_o = dbdd_o.attack()
+    assert (solution is not None) and (solution_o is not None)
+
```

Comment:
Can't we also test that both solutions are equal ?

Reply:
We can imagine that DBDD and DBDD_optimized don't integrate the same q-vectors. Since valid q-vectors are equivalent, it would give an equivalent DBDD instance, but it would not give the same solution because the output solution is `self.u` (which is `self.u_original` modified with `integrate_short_vector_hint` projections) and not `self.u_original` it-self. Initially, I inserted this condition, but for this reason, it removed it. In fact, I think this condition is not necessary because `dbdd(_o).attack()` already checks the solution (with `self.u`).

Reply:
:+1: